### DISCUSSION TITLE
post/naudio-20-year-anniversary

### DISCUSSION THREAD

**Discussion by giscus** [markheath/soundcode-blog-comments#1] [Announcements]
# post/naudio-20-year-anniversary

Mark Heath's Development Blog

https://markheath.net/post/naudio-20-year-anniversary

**Comment by cimpok** [comment]
Thank you for you work and for sharing this. I have not yet used your library yet, but I knew of its existence, and it was good to know that if I needed it for a project with audio, it is there.
How disheartening it should be for a novice programmer to read that a pretty successful project (with 4 million downloads it is) got here. What does he or she need to program then to make a living?

**Reply by markheath** [reply]
Yes, I don't think its particularly easy to monetise an open source project. However, building NAudio was a lot of fun and opened other doors of opportunity for me, so I'm not complaining.

**Comment by orwo1** [comment]
I have used NAudio and its been a pleasure, thank you for your contribution and efforts.
Are there modern alternatives you know of that might be a kind of replacement for NAudio?
If not now, maybe in the future once more developed?

**Reply by markheath** [reply]
There are a couple of other .NET audio libraries, although I'm not particularly familiar with them. The main alternative would be to use a native audio library and create .NET wrappers for it. It depends a bit on exactly what you're trying to achieve as to whether that's a viable option.

**Reply by orwo1** [reply]
Which libraries do you mean by "native audio library"?
Also, wrapping them mean support for a specific platform each, right?
Mostly I'm interested in playing music files like mp3 on Windows and possibly mobile platform like Xamarin.Forms(or maybe MAUI now?).

**Reply by markheath** [reply]
I mean something like PortAudio http://portaudio.com/
For just playing MP3, NAudio will work just great. You can also use the Audio Graph APIs https://learn.microsoft.com/en-us/windows/uwp/audio-video-camera/audio-graphs which are fine
I'm not too familiar with the audio choices on MAUI - I suspect that's limited at the moment

**Reply by tadadozo** [reply]
You did a fantastic job realizing naudio as it is right now.
Being very familiar with c# I couldn't find any better .net library and your library went far byoond my expectations.
Thanks for that, brilliant.
examples...
...last year we wanted to transform a live heartbeat from a polar bluetooth bealt into audio an light signals. Without naudio this wouldn't have been possible.
As a fallback I used the naudio library to simulate heartbeat sounds and extending your WaveProviders and your samples and documentation it was not that difficult. Thanks to your really breathtaking design I could concentrate on that what I wanted to realize.
It worked so good.


...between the years and the next art idea in mind (using multiple speakers reproducing daily scenarios in a room where people get surrounded by that scenario, thats nothing new, thats around already using 5.1, 7.1, 9.4.1, or Atmos or whatever)
But all that is not so easy. Its easy to consume, but not so easy to create.
So I tried to simply record my multi channel sacds,dvds or blu rays using the analogue way.
sounds easy, but...Next using sox to create a 5.1 flac and the sony x8000m2 can play it without burning dvd-audio, blu-ray, simply copy the result to a stick. For recording I used the outputs of an old cambridge audio bd650.
Thanks for all of your effort.
For your information:
AsioOut works for me perfectly with 6 channels, bytesPerSample=256 was no problem, Measured the time between the Audio_Available events and they are in a good 2-3ms range using the tascam 16x8 audio interface.

summary:
You have achieved so much for all of us out there, time is crucial I know, enjoy your time. It's so much fun using your library.
thanks lot.

**Reply by markheath** [reply]
wow, thanks @tadadozo, it's so cool to hear about all the things that NAudio has been used for. And also great to hear that you are successfully using ASIO. It is a little bit cumbersome to use in NAudio, but its always been one of the best ways to get low latency multi-channel audio playback on Windows.

**Comment by gmeisel01** [comment]
Mark,
Thank you for creating such a robust C# audio API!

I however need some help with figuring out the correct code example to get amplitude levels of each channel of an ASIO record feed so I can then translate it into GUI progress bars. 
Can you please help demystify how to extract the separate channel samples and then convert them into a usable float value?
